### PR TITLE
stdenv: support __structuredAttrs, opt-in per package

### PR DESCRIPTION
**Pull request by gnprice**
###### Motivation for this change

This PR is inspired by #72074, and is
 * a short, simple, and (I hope) mergeable PR,
 * that enables individual packages to start opting into structured attrs,
 * and also smooths the path to merging the bulk of the changes in #72074.

The key design choice in this PR is that we craft our stdenv changes to make it work both with and without structured attrs enabled.  Happily, this only adds a modest amount of complexity relative to supporting only the structured-attrs case.

This PR replaces only a small fraction of the changes in #72074.  The majority of the work represented in that PR is about fixing individual packages (lots and lots of them!) to work correctly with structured attrs.  My hope is that merging this PR will in turn make it easier to merge many of those changes too.

I'd love to hear feedback from the people who've been authoring and reviewing #72074, including @globin @lheckemann @jtojnar @Ericson2314 @FRidh @teto @Ma27, as well as naturally anyone else interested.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

I've tested this PR from two angles:

 * **Clean as is**: With this PR as is, not yet turning on structured attrs, I believe it has no effect on how any package is built.  (Modulo unknown bugs.)

 * **Works when enabled**: For most of the simpler packages in the tree, turning on structured attrs on top of this PR should Just Work.  That means no further edits are necessary, and again no effect on how the package gets built.

   (More-complex packages will need edits of their own in order to migrate; #72074 includes many examples.)

I've tested each of these on a random sample of packages.  To find subtle differences that don't simply cause the build to fail, I compared the build log against baseline.

For comparing build logs, a plain diff shows a lot of noise -- due to the different output path, and to nondeterminism in e.g. the order sources get compiled in.  I wrote a little script to normalize these out.  In many cases that made the diff between logs completely empty; in the rest, I read through the remaining diff by hand.

<details>

<summary>The results were (expand here):</summary>

 * **Clean as is**: I sampled 10 packages using `grep -wl stdenv.mkDerivation | shuf` (skipping 1 unfree).  I also took the same sample of 25 simpler packages used for the other test, below.

   35/35 built successfully.  26/35 (5/10 of the broad sample, 21/25 of the simple-package sample) had identical build logs after normalization; and for the other 9/35 I read the complete diffs and determined there was no change.  So 35/35 had no change detectable in the build log.

   One of these packages was `palemoon`, a web browser, with naturally a fairly complex build.

   Also naturally I rebuilt the whole toolchain at the tip of the PR, and everything seems to work fine.  I didn't inspect the build logs for this part.

 * **Works when enabled**: I sampled 25 packages from the simpler packages in the tree (direct callers of stdenv.mkDerivation, <= 30 lines long.)
    * 25/25 build successfully with structured attrs.
    * For 22/25, the build log shows no detectable difference.  (Of these 16 had empty diff after normalization; the other 6 were short and easy to read.)
    * 3/25 were more interesting:
      * For one package (`gbsplay`), the build logs expose a bug in the package expression!  I'll send the fix as its own small PR (#85044).  After fixing the bug, enabling structured attrs has no effect, as hoped.
      * For one package (`artha`), the log shows an extra "Nothing to be done" line from `make`.  Harmless in itself, but I plan to investigate in case this is a subtle sign of some bug.
      * For one package (`closurecompiler`), when structured attrs is enabled the calculation of `SOURCE_DATE_EPOCH` gets thrown off by the `.attrs.sh` file.  I plan to debug this and fix.

      I should emphasize that the effects on these 3/25 only appeared when structured attrs was actually enabled.  At the tip of the PR, with `__structuredAttrs` remaining false, I built these packages as part of the "clean as is" testing above, and confirmed the build logs remained unchanged (after normalization.)

</details>

In short:

 * As is, these changes left the build unaffected for 100% of packages I tried, including some complex ones.

 * On taking the further step of actually enabling structured attrs, the build remained unaffected for about 90% of a sample of relatively-simple packages.

### CODE REVIEW

**Comment by gnprice** [comment]
> I've tested each of these on a random sample of packages. To find subtle differences that don't simply cause the build to fail, I compared the build log against baseline.
> 
> For comparing build logs, a plain diff shows a lot of noise -- due to the different output path, and to nondeterminism in e.g. the order sources get compiled in. I wrote a little script to normalize these out. In many cases that made the diff between logs completely empty; in the rest, I read through the remaining diff by hand.

In case they're useful to anyone else, or if only to help make clear just how these tests were done, here's:
* the small [scaffold](https://gist.github.com/gnprice/ab0d9a6d97bec5c9a0f3813b8f57c8c9) I used for building each sampled package;
* the [`difflog` script](https://gist.github.com/gnprice/1864b871656773f78f419201e8b5cda3) I used for normalizing out boring changes in the build logs.

I'm pretty happy with the `difflog` tool -- it did quite a lot to make the diffs manageable to read. As mentioned in the details-expando in the OP, it made it possible for me to spot a couple of subtle (one- or two-line) changes when enabling structured attrs on certain packages. Both of those, I had completely missed when scanning through the same build logs before I went and wrote that tool.

The other essential ingredient in reading these logs was `git diff --no-index`, and in particular the "color-moved" feature of `git diff`. After normalization, some packages' log diffs remained several thousand lines long. With that feature [enabled](https://gist.github.com/gnprice/5f71b63464a76bd7e0480f207b1bbff3), almost every line was in a color showing it was merely reordered (in particular lots of compiler warnings), so I could scan through the whole thing and look closely at only the lines that had actually changed.

**Comment by Ericson2314** [comment]
I can get on board with this sort of thing, but I think it's important we have a bounded transitional period and don't go on supporting both indefinitely. For example I would say if we merge this now, commit to removing the old way by the release 20.09, and if the deadline comes before all the old packages are converted, too bad.

**Comment by gnprice** [comment]
> I can get on board with this sort of thing, but I think it's important we have a bounded transitional period and don't go on supporting both indefinitely.

Sure. From my perspective at least, supporting both is just a migration strategy — a means to help the migration go faster. I'd love to see it progress to being able to hardwire the `__structuredAttrs` in `mkDerivation` to true within a few months, and I'd be glad to contribute to making that happen if y'all / the community are on board with it.

**Comment by globin** [comment]
I'd prefer not to do this currently until the work on stdenv has been stabilised. Only yesterday I pushed more changes to that. Also `nix-shell -A` doesn't have support for strucutured-attrs builds, which I'm planning to look at in the next ~2 weeks. I don't think this makes any sense right now, sorry, maybe in 3-4 weeks but I hope when the larger issue with nix-shell is fixed we might just be able to merge it at once. And we should probably go through an RFC as we're changing `stdenv/mkDerivation` API in places.

**Comment by teto** [comment]
> a means to help the migration go faster. 

alternatively forcing the use of structuredAttrs can be a good way to incite people polish the code (assuming the first merge deals with most of the issues).  This may be an unpopular (or misinformed) opinion but I would be fine freezing unstable for a ~ month  while solving most issues. 
If done early enough (with some notice for people to postpone new PRs etc), we could avoid the qt  situation.

**Comment by Ericson2314** [comment]
Maybe the RFC could have both the "cold turkey" and "incremental migration with deadline" approaches, and the community could decide. If by some miracle everyone was on board with the former, I think that could be a really good exercise in community-wide coordination, not just a faster way to get things done.

**Comment by xaverdh** [comment]
What needs to be done to get this over the finish line?
I just read https://github.com/NixOS/nixpkgs/pull/41276 and was reminded of why it is a good idea to have this sooner rather than later..

**Comment by maisiliym** [comment]
> I'd prefer not to do this currently until the work on stdenv has been stabilised.

That makes sense; stabilize the interface before destabilizing it again.

**Comment by rnhmjoj** [comment]
> I'd prefer not to do this currently until the work on stdenv has been stabilised. Only yesterday I pushed more changes to that. Also nix-shell -A doesn't have support for strucutured-attrs builds, which I'm planning to look at in the next ~2 weeks. 

Has this been done yet?  Structured attrs are blocking the introduction of declarative wrappers, which are supposed to solve one of the major pains with packaging graphical applications: https://github.com/NixOS/nixpkgs/pull/83321#discussion_r397652024

> I don't think this makes any sense right now, sorry, maybe in 3-4 weeks

We have been waiting for years and packages are currently broken because we can't implement a solution without this. Can we *please* start to do something?

**Comment by globin** [comment]
This is a nix prerequisite before we can merge anything to master: https://github.com/NixOS/nix/pull/4770

Also RFCs for the stdenv changes are necessary

But what can definitely be done in the meantime is fix further packages and packaging infrastructure on this branch.

**Comment by Ma27** [comment]
I should note though that  NixOS/nix#4770 seems to only need a final approval from Eelco :) 

The other question is whether we can get this backported somehow, otherwise you'd need `nixUnstable` to properly debug packages using `__structuredAttrs = true;`.

**Comment by Shados** [comment]
`nix-shell` support for structured attrs made it into the Nix 2.4 release, is there anything else still blocking this...?

**Comment by Artturin** [comment]
I'm not aware of any blockers but at least the merge conflicts need to be solved

**Comment by Mindavi** [comment]
Is an RFC still needed to be able to get this in? I guess we do at least need a plan for when we want to move over to structuredAttrs (since it doesn't make sense to support both ways for an extended period of time).

**Comment by Artturin** [comment]
merge conflicts resolved in https://github.com/NixOS/nixpkgs/pull/175649
